util/admission: TestWorkQueueTokenResetRace failed

util/admission.TestWorkQueueTokenResetRace [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21133282?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21133282?buildTab=artifacts#/) on master @ [909a54745a560428a96a2c607e54e7a35e7fff86](https://github.com/cockroachdb/cockroach/commits/909a54745a560428a96a2c607e54e7a35e7fff86):


```
        	GOROOT/src/testing/testing.go:2477 +0x38
        testing.tRunner(0x4000216380, 0x4000259b08)
        	GOROOT/src/testing/testing.go:1934 +0xc8
        testing.runTests(0x400040a468, {0x24372a0, 0x22, 0x22}, {0x4000259ba8?, 0x4139a4?, 0x24d9e80?})
        	GOROOT/src/testing/testing.go:2475 +0x3b8
        testing.(*M).Run(0x40002fa3c0)
        	GOROOT/src/testing/testing.go:2337 +0x530
        main.main()
        	bazel-out/aarch64-fastbuild/bin/pkg/util/admission/admission_test_/testmain.go:208 +0x384
        goroutine 66 [chan receive]:
        github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
        	pkg/util/log/log_flush.go:123 +0x50
        created by github.com/cockroachdb/cockroach/pkg/util/log.init.5 in goroutine 1
        	pkg/util/log/log_flush.go:78 +0x24
        goroutine 67 [chan receive, 2 minutes]:
        github.com/cockroachdb/cockroach/pkg/util/log.signalFlusher()
        	pkg/util/log/log_flush.go:147 +0x70
        created by github.com/cockroachdb/cockroach/pkg/util/log.init.5 in goroutine 1
        	pkg/util/log/log_flush.go:79 +0x30
        goroutine 62 [syscall, 2 minutes]:
        os/signal.signal_recv()
        	GOROOT/src/runtime/sigqueue.go:152 +0x30
        os/signal.loop()
        	GOROOT/src/os/signal/signal_unix.go:23 +0x1c
        created by os/signal.Notify.func1.1 in goroutine 67
        	GOROOT/src/os/signal/signal.go:152 +0x28
        goroutine 5 [chan receive]:
        github.com/cockroachdb/cockroach/pkg/util/goschedstats.init.0.func1()
        	pkg/util/goschedstats/runnable_enabled.go:144 +0x8c
        created by github.com/cockroachdb/cockroach/pkg/util/goschedstats.init.0 in goroutine 1
        	pkg/util/goschedstats/runnable_enabled.go:136 +0x24
        goroutine 1077 [sleep]:
        time.Sleep(0x5f5e100)
        	GOROOT/src/runtime/time.go:363 +0x150
        github.com/cockroachdb/cockroach/pkg/util/admission.maybeRetryWithWait(0x400053cc40, {0x1640f51, 0x5}, 0x0, 0x400001b790)
        	pkg/util/admission/work_queue_test.go:377 +0x68
        github.com/cockroachdb/cockroach/pkg/util/admission.TestWorkQueueTokenResetRace.func1()
        	pkg/util/admission/work_queue_test.go:695 +0x274
        created by github.com/cockroachdb/cockroach/pkg/util/admission.TestWorkQueueTokenResetRace in goroutine 1061
        	pkg/util/admission/work_queue_test.go:663 +0x288: 
        Leaked goroutine: goroutine 1077 [sleep]:
        time.Sleep(0x5f5e100)
        	GOROOT/src/runtime/time.go:363 +0x150
        github.com/cockroachdb/cockroach/pkg/util/admission.maybeRetryWithWait(0x400053cc40, {0x1640f51, 0x5}, 0x0, 0x400001b790)
        	pkg/util/admission/work_queue_test.go:377 +0x68
        github.com/cockroachdb/cockroach/pkg/util/admission.TestWorkQueueTokenResetRace.func1()
        	pkg/util/admission/work_queue_test.go:695 +0x274
        created by github.com/cockroachdb/cockroach/pkg/util/admission.TestWorkQueueTokenResetRace in goroutine 1061
        	pkg/util/admission/work_queue_test.go:663 +0x288
--- FAIL: TestWorkQueueTokenResetRace (6.03s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/admission-control
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkQueueTokenResetRace.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-60576

--- a/pkg/util/admission/work_queue_test.go
+++ b/pkg/util/admission/work_queue_test.go
@@ -10,6 +10,7 @@ import (
 	"fmt"
 	"strconv"
 	"strings"
+	"sync"
 	"testing"
 	"time"
 	"unicode"
@@ -734,7 +735,8 @@ func TestWorkQueueTokenResetRace(t *testing.T) {
 	stopCh := make(chan struct{})
 	errCount, totalCount := 0, 0
 	var mu syncutil.Mutex
-	go func() {
+	var wg sync.WaitGroup
+	wg.Go(func() {
 		ticker := time.NewTicker(time.Microsecond * 100)
 		done := false
 		var work *testWork
@@ -745,7 +747,9 @@ func TestWorkQueueTokenResetRace(t *testing.T) {
 				ctx, cancel := context.WithCancel(context.Background())
 				work2 := &testWork{cancel: cancel}
 				tenantID++
+				wg.Add(1)
 				go func(ctx context.Context, tenantID uint64, createTime int64) {
+					defer wg.Done()
 					resp, err := q.Admit(ctx, WorkInfo{
 						TenantID:   roachpb.MustMakeTenantID(tenantID),
 						CreateTime: createTime,
@@ -786,7 +790,7 @@ func TestWorkQueueTokenResetRace(t *testing.T) {
 				buf.stringAndReset()
 			}
 		}
-	}()
+	})
 	go func() {
 		for {
 			select {
@@ -803,6 +807,7 @@ func TestWorkQueueTokenResetRace(t *testing.T) {
 	time.Sleep(time.Second)
 	close(stopCh)
 	q.close()
+	wg.Wait()
 	mu.Lock()
 	t.Logf("total: %d, err: %d", totalCount, errCount)
 	mu.Unlock()

